### ISSUE TITLE
False positive cppcoreguidelines-virtual-class-destructor warning for templated class

### ISSUE THREAD

**Issue by 934344f2-d638-41f4-be47-07f2d2914bec** [llvm/llvm-project#51759]
|  |  |
| --- | --- |
| Bugzilla Link | [52417](https://llvm.org/bz52417) |
| Version | unspecified |
| OS | All |
| CC | @EugeneZelenko,@fabianwolff |

## Extended Description 
When a derived class is templated, the destructor will not be recognized by clang-tidy and the cppcoreguidelines-virtual-class-destructor warning is issued. See this example:

class base {
public:
  virtual ~base() noexcept;
};

template <typename T>
class derived : public base {
public:
  ~derived() noexcept override;
};

void test() {
  derived<int> a;
}

**Comment by captaincrutches** [comment]
I'm able to reproduce this on clang 14 with 2 levels of template inheritance - that is, a base class; a derived class template; and a second class template derived from the first:
```cpp
class Base
{
public:
    virtual ~Base() = default;
};

template <typename T>
class Derived1 : public Base
{
public:
    ~Derived1() override = default;
};

template <typename T>
class Derived2 : public Derived1<T>
{
public:
    ~Derived2() override = default;
};
```
clang-tidy tells me:
```
error: destructor of 'Derived2' is public and non-virtual [cppcoreguidelines-virtual-class-destructor,-warnings-as-errors]
class Derived2 : public Derived1<T>
      ^
```
Notably, it does _not_ happen with only `Derived1`, and it does _not_ happen if `Base` is also a template.

**Comment by PiotrZSL** [comment]
Looks to be fixed with f0711106dc6c14dcaf06437a0467043e983bf9dc
